tools/msbuild-fuzzer/Program.cs

```diff
@@ -0,0 +1,311 @@
+			return builder.RunTarget (application, "Restore", doNotCleanupOnUpdate: true);
+		}
+
+		static bool Build ()
+		{
+			Console.WriteLine (nameof (Build));
+			return builder.Build (application, doNotCleanupOnUpdate: true);
+		}
+
+		static bool DesignTimeBuild ()
+		{
+			Console.WriteLine (nameof (DesignTimeBuild));
+			return builder.DesignTimeBuild (application, doNotCleanupOnUpdate: true);
+		}
+
+		static readonly string [] DesignerParameters = new [] { "DesignTimeBuild=True", "AndroidUseManagedDesignTimeResourceGenerator=False" };
+
+		static bool DesignerBuild ()
+		{
+			Console.WriteLine (nameof (DesignerBuild));
+			return builder.RunTarget (application, "SetupDependenciesForDesigner", doNotCleanupOnUpdate: true, parameters: DesignerParameters);
+		}
+
+		static bool Clean ()
+		{
+			Console.WriteLine (nameof (Clean));
+			return builder.Clean (application, doNotCleanupOnUpdate: true);
+		}
+
+		static bool Install ()
+		{
+			Console.WriteLine (nameof (Install));
+			if (!builder.Install (application, doNotCleanupOnUpdate: true)) {
+				return false;
+			}
+			if (!installedPackages.Contains (application.PackageName))
+				installedPackages.Add (application.PackageName);
+			needsInstall = false;
+			return true;
+		}
+
+		static bool Uninstall ()
+		{
+			foreach (var packageName in installedPackages) {
+				Adb ($"uninstall {packageName}");
+			}
+			installedPackages.Clear ();
+			needsInstall = true;
+			return true;
+		}
+
+		static bool Run ()
+		{
+			if (needsInstall && !Install ())
+				return false;
+			Console.WriteLine (nameof (Run));
+			string stop = Adb ($"shell am force-stop {application.PackageName}", ignoreExitCode: true);
+			Adb ("logcat -c");
+			string activity = $"{application.PackageName}/md52d9cf6333b8e95e8683a477bc589eda5.MainActivity";
+			string start = Adb ($"shell am start -n {activity}");
+			Console.WriteLine (start.Trim ());
+			//Wait for the app to start
+			Thread.Sleep (3000);
+			string logcat = Adb ("logcat -d");
+			if (!logcat.Contains ("App started!")) {
+				Console.WriteLine (logcat);
+				throw new Exception ($"Activity {activity} did not start!");
+			}
+			return true;
+		}
+
+		static string Adb (string arguments, bool ignoreExitCode = false)
+		{
+			var info = new ProcessStartInfo {
+				FileName = "adb",
+				Arguments = arguments,
+				CreateNoWindow = true,
+				WindowStyle = ProcessWindowStyle.Hidden,
+				UseShellExecute = false,
+				RedirectStandardError = true,
+				RedirectStandardOutput = true,
+			};
+			var builder = new StringBuilder ();
+			using (var process = new Process ()) {
+				process.StartInfo = info;
+				process.OutputDataReceived += (sender, e) => builder.AppendLine (e.Data);
+				process.ErrorDataReceived += (sender, e) => builder.AppendLine (e.Data);
+				process.Start ();
+				process.BeginErrorReadLine ();
+				process.BeginOutputReadLine ();
+				process.WaitForExit ();
+
+				if (!ignoreExitCode && process.ExitCode != 0) {
```

Comment:
Should we be using the `ManualResetEvent` stuff we have in https://github.com/xamarin/xamarin-android/blob/master/build-tools/Xamarin.Android.Tools.BootstrapTasks/Xamarin.Android.Tools.BootstrapTasks/Adb.cs#L217. Or are we sure we won't get any output after the `WaitForExit` call has completed?